src/metrics/tagfiltertree/tag_filter_tree.go

```diff
@@ -0,0 +1,317 @@
+			Name: tag.Name,
+		})
+		nodeIdx = len(t.Nodes) - 1
+	}
+	node := t.Nodes[nodeIdx]
+
+	// Add the data to the childTree if this is the last tag.
+	var dataToAdd *T
+	if idx == len(tags)-1 {
+		dataToAdd = &data
+	}
+
+	childTree, err := node.addValue(tag.Val, dataToAdd)
+	if err != nil {
+		return err
+	}
+
+	// Recurse to the next tag.
+	if err := addNode(childTree, tags, idx+1, data); err != nil {
+		// TODO: perform cleanup to avoid partially added nodes.
+		return err
+	}
+
+	return nil
+}
+
+func match[T any](
+	t *Tree[T],
+	tags map[string]string,
+	data *[]T,
+	isMatchAny bool,
+) (bool, error) {
+	if len(tags) == 0 || t == nil {
+		return false, nil
+	}
+
+	for _, node := range t.Nodes {
+		name := node.Name
+		negate := false
+		if IsMatchNoneTag(name) {
+			name = name[1:]
+			negate = true
+		}
+		tagValue, tagNameFound := tags[name]
+		absVal, absValFound := node.AbsoluteValues[tagValue]
+		if tagNameFound && absValFound {
+			if data != nil && *data != nil {
+				*data = append(*data, absVal.Data...)
+			}
+			if isMatchAny && len(absVal.Data) > 0 {
+				return true, nil
+			}
+
+			matched, err := match(absVal.Tree, tags, data, isMatchAny)
+			if err != nil {
+				return false, err
+			}
+			if isMatchAny && matched {
+				return true, nil
+			}
+		}
+		if tagNameFound != negate {
+			for _, v := range node.PatternValues {
+				matched, err := v.PatternTrie.Match(tagValue, nil)
+				if err != nil {
+					return false, err
+				}
+				if matched {
+					if data != nil && *data != nil {
+						*data = append(*data, v.Data...)
+					}
+					if isMatchAny && len(v.Data) > 0 {
+						return true, nil
+					}
+					matched, err := match(v.Tree, tags, data, isMatchAny)
+					if err != nil {
+						return false, err
+					}
+					if isMatchAny && matched {
+						return true, nil
+					}
+				}
+			}
+		}
+	}
+
+	return false, nil
+}
+
+// TagsFromTagFilter creates tags from a tag filter.
+// The tag values can be of the format:
+// "foo" OR "{foo,bar,baz}" OR "{{Variable}}"
+// There cannot be a mix of value formats like "simpleValue, {foo,bar}" etc.
+func TagsFromTagFilter(tf string) ([]Tag, error) {
+	tags := make([]Tag, 0)
+	tfSanitized := strings.TrimSpace(tf)
+	tagValuePairs := strings.Split(tfSanitized, " ")
+	for _, tagValuePair := range tagValuePairs {
+		tagAndValue := strings.Split(tagValuePair, ":")
+		if len(tagAndValue) != 2 {
+			return nil, fmt.Errorf("invalid tag filter: %s", tf)
+		}
+		tag := tagAndValue[0]
+		val := tagAndValue[1]
+		if len(tag) == 0 || len(val) == 0 {
+			return nil, fmt.Errorf("invalid tag filter: %s", tf)
+		}
+
+		varName := ""
+		if IsVarTagValue(val) {
+			varName = val
+			val = _matchall
+		}
+		if val == _matchNone {
+			tag = "!" + tag
+			val = _matchall
+		}
+
+		tags = append(tags, Tag{
+			Name: tag,
+			Val:  val,
+			Var:  varName,
+		})
+	}
+
+	return tags, nil
+}
+
+// IsVarTagValue returns true if the value is a variable tag value.
+func IsVarTagValue(value string) bool {
```

Comment:
add some explanation? e.g. I am not sure why len<4 means false

Reply:
a Var value type is of the form "{{...}}" so the minimum is 4 chars. Anything lesser and we know it's not a Var value type

Reply:
nit: add comments to explain what Var Tag means. No reviewer can correlate "VarTagValue" with "...{{...}}..."